### DISCUSSION TITLE
Until zfs native encryption becomes fast and stable, use LUKS for local encryption and restic for encrypted incremental backups.

### DISCUSSION THREAD

**Discussion by amano-kenji** [openzfs/zfs#16589] [Ideas]
Right now, zfs native encryption is slower than LUKS. It is also unstable and leads to kernel panics and various errors when sending or receiving raw encrypted zfs snapshots.

Thus, I propose LUKS for local encryption and restic for making encrypted incremental backups.

Restic stores file blocks and makes incremental snapshots. It can delete arbitrary restic snapshots without losing data just like zfs snapshots. It supports compression, deduplication, and encryption. Thus, you can store restic snapshots on untrusted remote machines. You can use amazon cloud storage, https://rsync.net/, https://zfs.rent/, and other cloud storage providers.

Restic 0.17 started supporting `RESTIC_FEATURES=device-id-for-hardlinks` which supports backing up $ZFS-MOUNTPOINT/.zfs/snapshot/$SNAPSHOT-NAME efficiently. Restic 0.18 will remove device-id-for-hardlinks feature flag and support .zfs/snapshot directories efficiently witout any feature flag.

If you want to back up zfs dataset, you can take `restic` ZFS snapshot, back up .zfs/snapshot/restic as a new restic snapshot, and delete restic ZFS snapshot after backing it up. In this way, restic doesn't need to know about local sanoid ZFS snapshots which are independent from restic snapshots.

**Comment by amotin** [comment]
Do you plan to pay us for the advertisement of your products while blaming ZFS for problems with no details?

**Reply by amano-kenji** [reply]
I don't sell any product. And, LUKS and restic are free and open source. Nobody gets paid for this. If I wanted to be paid, I would probably offer affiliate links to cloud services. @rincebrain has documented the problems with zfs native encryption which is essentially unmaintained at this point.

I really wanted to use zfs native encryption, but it is likely to lead to kernel panics and various errors according to various github issues when I send raw encrypted zfs snapshots.

https://github.com/openzfs/openzfs-docs/issues/494 demands that zfs should warn people about zfs native encryption especially when they send or receive raw encrypted zfs snapshots.

Until zfs native encryption is fixed, people need to know alterantives to zfs native encryption.

There is no blame for zfs. The fact that zfs native encryption is broken at the moment is well documented on this github repository. Is speaking truth a sin? You hate me for speaking truth? While zfs native encryption is probably fine for local usage, people should be warned that it is broken for encrypted backups.

**Reply by amotin** [reply]
Speaking truth can be constructive or destructive.  Constructive includes links to still open issues and other means to solve the problem.  Destructive just throws acquisitions like "slower" and "unstable and leads to kernel panics and various errors", which do not lead to any improvement, but only a disrespect to developers working on the project and a discouragement for users.  I personally fixed several encryption issues few months ago, while I don't even use encryption myself at this point, and I'd appreciate some real feedback on my work instead of this.

**Reply by rincebrain** [reply]
I think people are frustrated because the perception is that the project should be indicating there are a good number of outstanding issues with it in a similar way to the warnings about dedup, at a minimum, and the stonewalling on the topic leads to a number of people getting burned and then being even more frustrated when the response to their asking for help is "yeah, you should have asked people before using it, we know it's busted".

Perhaps we should listen to the people shouting that they want us to have a warning that we have this rough edge in the project.

**Reply by amano-kenji** [reply]
> I'd appreciate some real feedback on my work instead of this.

I didn't come here to give a feedback to your work. I don't know developers individually.

If you want compliments, you should advertise your efforts widely on the internet instead of being mad at someone else's discussion topic and closing it.

If you want fame, john carmack is your role model. He at least advertises his efforts.

Despite your efforts, people still report kernel panics over sending or receiving raw encrypted zfs snapshots.

For example, https://github.com/openzfs/zfs/issues/11679

Because you exerted efforts on zfs encryption, discussion on instability of zfs encryption should be restricted? That's bad. Don't be selfish. At least, give people some warning about zfs encryption. People should be informed so that they know what they are getting into.

I don't criticize your efforts, but I criticize you for selfishly hiding information that can help other people.

There is a lot of work to be done before zfs encryption doesn't cause major issues. Until we get there, let's be honest about the progress. Zfs encryption is not mature, yet. Don't give people any impression that zfs encrypted snapshots are ready for production.

**Reply by rincebrain** [reply]
Please, let's play nice in the sandbox, everyone.

I'm not thrilled about where this part of the project is either, as discussions have previously shown at length, but let's go for civil conversation.

**Reply by amano-kenji** [reply]
https://github.com/openzfs/zfs/issues/12014 still reports that zfs encryption corrupts snapshots. Zfs native encryption isn't ready for production.

**Reply by lundman** [reply]
As someone who has used zfs native encryption for more than 10 years, and not once had issues with it, it is working very well. It plugs in the same as compression after all.
Now send/receive on the other hand has had many problems; remember -L anyone? Incremental send holes? Yep. I'm here to champion for renaming the issues from "native encryption is broken" to "send/recv is broken". All hail hyd^Wencryption!

**Reply by rincebrain** [reply]
Jorgen, if send/recv works reliably when not using native encryption, and if some of the bugs in native encryption can be triggered without using send/recv, the correct overarching description for the data points does not seem to be "send/recv is buggy".

**Reply by amano-kenji** [reply]
There is a possibility that some of those issues were caused by a lot of dust in computer case. I thought I had to replace RAM or GPU because they were problematic. After thoroughly eliminating dust in my computer case, errors in RAM and GPU went away.

I think dust is charged with electrons that zap computer parts.

Many errors will go away if you replace non-ECC RAM with ECC RAM or remove dust in your computer case.

**Comment by vuongtuha** [comment]
What the heck is kernel panic with native encryption? Ping me issue pls
Btw i used borg backup only, am i kicked out?

**Reply by IvanVolosyuk** [reply]
Chill, fam. Just glow up your ZFS to 2.2.8 or 2.3.3 and then send it with 'zpool scrub'. https://github.com/openzfs/zfs/issues/12014
